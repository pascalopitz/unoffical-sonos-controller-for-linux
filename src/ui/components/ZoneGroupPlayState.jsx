import React from 'react';

export function ZoneGroupPlayState(props) {
    const { currentTrack, playState, mediaInfo } = props;

    const icon =
        playState === 'playing' ? (
            <i className="material-icons">play_arrow</i>
        ) : (
            <i className="material-icons">pause</i>
        );

    let info;

    if (currentTrack && currentTrack.title) {
        info = currentTrack.title;

        if (currentTrack.artist) {
            info = info + ' - ' + currentTrack.artist;
        }
    } else {
        info = mediaInfo?.CurrentURIMetaData?.['dc:title'];
    }

    if (info) {
        return (
            <div className="play-state">
                {icon} {info}
            </div>
        );
    }

    return <div className="play-state" />;
}

export default ZoneGroupPlayState;
