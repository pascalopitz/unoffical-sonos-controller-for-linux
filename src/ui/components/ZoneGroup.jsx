import React from 'react';
import { connect } from 'react-redux';

import ZoneGroupPlayState from './ZoneGroupPlayState';
import ZoneGroupMember from './ZoneGroupMember';

const { showGroupManagement, selectGroup } = window.ZoneGroupActions;

const mapDispatchToProps = (dispatch) => {
    return {
        showManagement: (group) => dispatch(showGroupManagement(group)),
        selectGroup: (group) => dispatch(selectGroup(group)),
    };
};

export function ZoneGroup(props) {
    const {
        group,
        showManagement,
        selectGroup,
        playStates,
        mediaInfos,
        currentTracks,
        currentHost,
    } = props;

    const _onClick = () => {
        selectGroup(group);
    };

    const _showGroupManagement = (e) => {
        showManagement(group);
        e.preventDefault();
        e.stopPropagation();
    };

    const playState = playStates[group.host] || 'stopped';
    const mediaInfo = mediaInfos[group.host] || {};
    const currentTrack = currentTracks[group.host] || {};

    const zoneNodes = group.ZoneGroupMember.map((item, index) => (
        <ZoneGroupMember member={item} key={index} />
    ));

    let classString = 'not-selected';

    if (currentHost && group.host === currentHost) {
        classString = 'selected';
    }

    classString += ' zone-group';

    return (
        <div className={classString} onClick={_onClick}>
            <ul>{zoneNodes}</ul>
            <div className="group-button" onClick={_showGroupManagement}>
                Group
            </div>
            <ZoneGroupPlayState
                mediaInfo={mediaInfo}
                playState={playState}
                currentTrack={currentTrack}
            />
        </div>
    );
}

export default connect(null, mapDispatchToProps)(ZoneGroup);
