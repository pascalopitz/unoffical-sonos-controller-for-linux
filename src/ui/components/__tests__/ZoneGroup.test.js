import React from 'react';
import { ZoneGroup } from '../ZoneGroup';
import ZoneGroupPlayState from '../ZoneGroupPlayState';
import { mount } from 'enzyme';

jest.mock('../ZoneGroupMember', () => () => <p className="member" />); // eslint-disable-line

jest.mock('../ZoneGroupPlayState');
ZoneGroupPlayState.mockReturnValue(<p className="play-state" />);

describe('ZoneGroup', () => {
    let props;

    beforeEach(() => {
        props = {
            group: {
                host: 'myhost',
                ZoneGroupMember: [
                    {
                        member: 1,
                        Coordinator: 'true',
                    },
                ],
            },
            selectGroup: jest.fn(),
            showManagement: jest.fn(),
            playStates: {
                myhost: 'playing',
            },
            mediaInfos: {
                myhost: {},
            },
            currentTracks: {
                myhost: {},
            },
        };
    });

    it('matches snapshot', () => {
        const context = mount(<ZoneGroup {...props} />);
        expect(context).toMatchSnapshot();
        expect(context.find('.not-selected').length).toBe(1);
        expect(ZoneGroupPlayState).toHaveBeenCalled();

        const [lastCall] = ZoneGroupPlayState.mock.calls;
        expect(lastCall[0]).toMatchObject({
            playState: 'playing',
        });

        context.unmount();
    });

    it('matches snapshot when selected', () => {
        const currentHost = 'myhost';

        props = {
            ...props,
            currentHost,
        };

        const context = mount(<ZoneGroup {...props} />);
        expect(context).toMatchSnapshot();
        expect(context.find('.selected').length).toBe(1);
        context.unmount();
    });

    it('clicking whole group selects it', () => {
        const context = mount(<ZoneGroup {...props} />);
        context.find('.zone-group').simulate('click');
        expect(props.selectGroup).toHaveBeenCalled();
        context.unmount();
    });

    it('clicking button opens management', () => {
        const context = mount(<ZoneGroup {...props} />);

        context.find('.group-button').simulate('click', {
            preventDefault: jest.fn(),
            stopPropagation: jest.fn(),
        });

        expect(props.showManagement).toHaveBeenCalled();
        context.unmount();
    });
});
