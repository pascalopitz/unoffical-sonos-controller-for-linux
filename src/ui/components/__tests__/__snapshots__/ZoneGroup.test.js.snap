// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoneGroup matches snapshot 1`] = `
<ZoneGroup
  currentTracks={
    Object {
      "myhost": Object {},
    }
  }
  group={
    Object {
      "ZoneGroupMember": Array [
        Object {
          "Coordinator": "true",
          "member": 1,
        },
      ],
      "host": "myhost",
    }
  }
  mediaInfos={
    Object {
      "myhost": Object {},
    }
  }
  playStates={
    Object {
      "myhost": "playing",
    }
  }
  selectGroup={[MockFunction]}
  showManagement={[MockFunction]}
>
  <div
    className="not-selected zone-group"
    onClick={[Function]}
  >
    <ul>
      <Component
        key="0"
        member={
          Object {
            "Coordinator": "true",
            "member": 1,
          }
        }
      >
        <p
          className="member"
        />
      </Component>
    </ul>
    <div
      className="group-button"
      onClick={[Function]}
    >
      Group
    </div>
    <ZoneGroupPlayState
      currentTrack={Object {}}
      mediaInfo={Object {}}
      playState="playing"
    >
      <p
        className="play-state"
      />
    </ZoneGroupPlayState>
  </div>
</ZoneGroup>
`;

exports[`ZoneGroup matches snapshot when selected 1`] = `
<ZoneGroup
  currentHost="myhost"
  currentTracks={
    Object {
      "myhost": Object {},
    }
  }
  group={
    Object {
      "ZoneGroupMember": Array [
        Object {
          "Coordinator": "true",
          "member": 1,
        },
      ],
      "host": "myhost",
    }
  }
  mediaInfos={
    Object {
      "myhost": Object {},
    }
  }
  playStates={
    Object {
      "myhost": "playing",
    }
  }
  selectGroup={[MockFunction]}
  showManagement={[MockFunction]}
>
  <div
    className="selected zone-group"
    onClick={[Function]}
  >
    <ul>
      <Component
        key="0"
        member={
          Object {
            "Coordinator": "true",
            "member": 1,
          }
        }
      >
        <p
          className="member"
        />
      </Component>
    </ul>
    <div
      className="group-button"
      onClick={[Function]}
    >
      Group
    </div>
    <ZoneGroupPlayState
      currentTrack={Object {}}
      mediaInfo={Object {}}
      playState="playing"
    >
      <p
        className="play-state"
      />
    </ZoneGroupPlayState>
  </div>
</ZoneGroup>
`;
